proposals/continuations/Explainer.md

```diff
@@ -0,0 +1,1277 @@
+The second way to invoke a continuation is to raise an exception at
+the control tag invocation site. This amounts to performing "an
+abortive action" which causes the stack to be unwound.
+
+
+```wat
+  resume_throw $exn : [tp* (ref $ct)])] -> [tr*]
+  where:
+  - $ct = cont ([ta*] -> [tr*])
+  - $exn : [tp*] -> []
+```
+
+The instruction `resume_throw` is parameterised by the exception to be
+raised at the control tag invocation site. As with `resume`, this
+instruction also fully consumes its continuation
+argument. Operationally, this instruction raises the exception `$exn`
+with parameters of type `tp*` at the control tag invocation point in
+the context of the supplied continuation. As an exception is being
+raised (the continuation is not actually being supplied a value) the
+parameter types for the continuation `ta*` are unconstrained.
+
+### Suspending Continuations
+
+A computation running inside a continuation can suspend itself by
+invoking one of the declared control tags.
+
+
+```wat
+  suspend $e : [tp*] -> [tr*]
+  where:
+  - $e : [tp*] -> [tr*]
+```
+
+The instruction `suspend` invokes the control tag named `$e` with
+arguments of types `tp*`. Operationally, the instruction transfers
+control out of the continuation to the nearest enclosing handler for
+`$e`. This behaviour is similar to how raising an exception transfers
+control to the nearest exception handler that handles the
+exception. The key difference is that the continuation at the
+suspension point expects to be resumed later with arguments of types
+`tr*`.
+
+### Binding Continuations
+
+The domain of a continuation may be shrunk via `cont.bind`. This
+instruction provides a way to partially apply a given
+continuation. This facility turns out to be important in practice due
+to the block and type structure of Wasm as in order to return a
+continuation from a block, all branches within the block must agree on
+the type of continuation. By using `cont.bind`, one can
+programmatically ensure that the branches within a block each return a
+continuation with compatible type (the [Examples](#examples) section
+provides several example usages of `cont.bind`).
+
+
+```wat
+  cont.bind $ct2 : [tp1* (ref $ct1)] -> [(ref $ct2)]
+  where:
+  $ct1 = cont ([tp1 tp2*] -> [tr*])
+  $ct2 = cont ([tp2*] -> [tr*])
+```
+
+The instruction `cont.bind` binds the arguments of type `tp1*` to a
+continuation of type `$ct1`, yielding a modified continuation of type
+`$ct2` which expects fewer arguments. This instruction also consumes
+its continuation argument, and yields a new continuation that can be
+supplied to either `resume`,`resume_throw`, or `cont.bind`.
+
+### Trapping Continuations
+
+In order to ensure that control cannot be captured across language
+boundaries, we provide an instruction for explicitly trapping attempts
+at reifying stacks across language boundaries.
+
+```wat
+  barrier $label $bt instr* : [s*] -> [t*]
+  where:
+  - $bt = [s*] -> [t*]
+  - instr* : [s*] -> [t*]
```

Comment:
```suggestion
  barrier $l bt instr* end : [t1*] -> [t2*]
  where:
  - bt = [t1*] -> [t2*]
  - instr* : [t1*] -> [t2*]
```
(Because a block type is not a type index but something else.)